File: javascript/two-pointers/3sum.js
var threeSum = function(nums) {

  // instructions
  // return unique arrays with values that add up to 0 for example [0,0,0] and [-3,2,1]
  // tip sort the array to make this easier

  // initialize an empty array to store the result
  const res = [];

  // sort the input array in ascending order
  nums.sort((a, b) => a - b);

  // initialize the index i to 0
  let i = 0;
  // iterate through the elements in the array
  while (i < nums.length) {
    // get the current element a at index i
    const a = nums[i];

    // check for duplicate elements, skip if duplicate
    if (i > 0 && a === nums[i - 1]) {
      i++;
      continue;
    }

    // initialize two pointers, l and r
    let l = i + 1;
    let r = nums.length - 1;

    // use a two-pointer approach to find triplets
    while (l < r) {
      const threeSum = a + nums[l] + nums[r];
      if (threeSum > 0) {
        r--;
      } else if (threeSum < 0) {
        l++;
      } else {
        // if a triplet is found, push it to the result array
        res.push([a, nums[l], nums[r]]);
        l++;
        // skip duplicate elements in the inner loop
        while (l < r && nums[l] === nums[l - 1]) {
          l++;
        }
      }
    }

    // increment the outer loop index i
    i++;
  }

  // return the final result array
  return res;
};

console.log(threeSum([-1, 0, 1, 2, -1, -4]));
console.log(threeSum([0, 1, 1]));
console.log(threeSum([0, 0, 0]));

File: python/neetcode/two-pointers/3sum.py
from typing import List

def threeSum(nums: List[int]) -> List[List[int]]:
    # initialize an empty list to store the result
    result = []
    
    # sort the input list in ascending order
    nums.sort()

    # initialize the index i to 0
    i = 0
    # iterate through the elements in the list
    while i < len(nums):
        # get the current element a at index i
        a = nums[i]

        # check for duplicate elements, skip if duplicate
        if i > 0 and a == nums[i - 1]:
            i += 1
            continue

        # initialize two pointers, l and r
        l = i + 1
        r = len(nums) - 1

        # use a two-pointer approach to find triplets
        while l < r:
            threeSum = a + nums[l] + nums[r]
            if threeSum > 0:
                r -= 1
            elif threeSum < 0:
                l += 1
            else:
                # if a triplet is found, append it to the result
                result.append([a, nums[l], nums[r]])
                l += 1
                # skip duplicate elements in the inner loop
                while l < r and nums[l] == nums[l - 1]:
                    l += 1

        # increment the outer loop index i
        i += 1

    # return the final result
    return result

print(threeSum([-1,0,1,2,-1,-4]))
print(threeSum([0,1,1]))
print(threeSum([0,0,0]))